OpenRA.Platforms.Default/OpenAlSoundEngine.cs

```diff
@@ -282,25 +282,33 @@ public void PauseSound(ISound sound, bool paused)
 				return;
 
 			var source = ((OpenAlSound)sound).Source;
-			int state;
-			AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
-			if (state == AL10.AL_PLAYING && paused)
-				AL10.alSourcePause(source);
-			else if (state == AL10.AL_PAUSED && !paused)
-				AL10.alSourcePlay(source);
+			PauseSound(source, paused);
 		}
 
 		public void SetAllSoundsPaused(bool paused)
 		{
 			foreach (var source in sourcePool.Keys)
+				PauseSound(source, paused);
+		}
+
+		void PauseSound(uint source, bool paused)
+		{
+			int state;
+			AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
+			if (paused)
 			{
-				int state;
-				AL10.alGetSourcei(source, AL10.AL_SOURCE_STATE, out state);
-				if (state == AL10.AL_PLAYING && paused)
+				if (state == AL10.AL_PLAYING)
 					AL10.alSourcePause(source);
-				else if (state == AL10.AL_PAUSED && !paused)
+				else if (state == AL10.AL_INITIAL)
+				{
+					// If a sound hasn't started yet,
+					// we indicate it should not play be transitioning it to the stopped state.
```

Comment:
This doesn't seem to work.  If I switch songs and immediately pause then the button in the music player will show the "play" state (meaning that it thinks the song is paused) but it will start to play as normal.  The button must be pressed twice to actually pause it, consistent with the player thinking that the song is indeed paused.

Reply:
Works on my machine just fine. :/

Can you put some breakpoints in:
- After we pause the unstarted audio by doing the play-stop, query the state and check if it is indeed `AL_STOPPED` or not.
- When it wants to start the audio, it checks the state before deciding to play or rewind it. What state does it report there? Again we are hoping to see `AL_STOPPED`.

Even if we can't fix this, do you think we can take it anyway? In theory the only way to replicate this would be to pause a track in the window where it's loading but hasn't started yet, which I don't imagine will happen too often. Plus it can be resolved with some additional clicking.

Reply:
On my machine trying to play/stop an `AL_INITIAL` source does not change the state.  It remains as `AL_INITIAL`.

I may be missing something, but shouldn't the fix be simple?  `OpenAlSound` already defines a `Stop` method, so we can add a `Pause` method that can set our own flag in the async sound that we can then check when it is ready to start, instead of relying on the internal `AL_SOURCE_STATE`.

Reply:
Its understandable if you want to see the backside of this feature, so let me know if you'd like me to do that fix.

Reply:
We then need to check that flag within the task, which incurs thread-safety concerns. This makes all the signalling required very messy very quickly. We need to have the task wait if paused before it's loaded, but resume once unpaused, and all that has to be done in a thread-safe way and also support cancellation when stopped.

Instead, I've pushed a change where I just attach a fake buffer on creation, swapping out for the real one later. I'm hoping attaching a buffer then allows the play/stop to actually enact the state change we're after. Sucky that we need the workaround, but it's far less mentally taxing.